jams/sonify.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+# CREATED:2015-12-12 18:20:37 by Brian McFee <brian.mcfee@nyu.edu>
+r'''
+Sonification
+============
+
+.. autosummary::
+    :toctree: generated/
+
+    sonify
+'''
+
+import six
+import numpy as np
+import mir_eval.sonify
+from mir_eval.util import filter_kwargs
+from .eval import coerce_annotation
+from .exceptions import NamespaceError
+
+__all__ = ['sonify']
+
+
+
+def clicks(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify clicks timings
+
+    '''
+
+    interval, _ = annotation.data.to_interval_values()
+
+    return filter_kwargs(mir_eval.sonify.clicks, interval[:, 0],
+                         fs=sr, length=length, **kwargs)
+
+
+def chord(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify chords'''
+
+    intervals, chords = annotation.data.to_interval_values()
+
+    return filter_kwargs(mir_eval.sonify.chords,
+                         chords, intervals,
+                         fs=sr, length=length,
+                         **kwargs)
+
+
+def pitch_hz(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify pitches in Hz'''
+
+    intervals, pitches = annotation.data.to_interval_values()
+
+    # Collapse down to a unique set of frequency values
+    freqs = np.unique(pitches)
+
+    # Build the piano roll
+    pitch_index = {p: i for i, p in enumerate(freqs)}
+    gram = np.zeros((len(freqs), len(pitches)))
+    for t, n in enumerate(pitches):
+        gram[pitch_index[n], t] = 1.0
+
+    return filter_kwargs(mir_eval.sonify.time_frequency,
+                         gram, freqs, intervals,
+                         fs=sr, length=length,
+                         **kwargs)
+
+
+SONIFY_MAPPING = {'beat': clicks,
+                  'segment_open': clicks,
+                  'onset': clicks,
+                  'chord': chord,
+                  'pitch_hz': pitch_hz}
+
+
+def sonify(annotation, sr=22050, duration=None, **kwargs):
+    '''Sonify a jams annotation through mir_eval
+
+    Parameters
+    ----------
+    annotation : jams.Annotation
+        The annotation to sonify
+
+    sr = : int > 0
+        The sampling rate of the output waveform
+
+    duration : float (optional)
+        Optional length (in seconds) of the output waveform
+
+    kwargs
+        Additional keyword arguments to mir_eval.sonify functions
+
+    Returns
+    -------
+    y_sonified : np.ndarray
+        The waveform of the sonified annotation
+
+    Raises
+    ------
+    NamespaceError
+        If the annotation has an un-sonifiable namespace
+    '''
+
+    length = None
+    if duration is not None:
+        length = int(duration * sr)
+
+    for namespace, func in six.iteritems(SONIFY_MAPPING):
+        try:
+            coerce_annotation(annotation, namespace)
+            return func(annotation, sr=sr, length=length, **kwargs)
+        except NamespaceError:
+            pass
```

Comment:
> a hypothetical future case where you want to have different sonification functions for two annotations that can be converted into each other

If two namespaces can be converted to each other, then they must be "equivalent" in a pretty strong sense.  It's hard to imagine why we would have different sonification methods there.

However, I think I can imagine a slightly more realistic case in which conversion only works in one way.  We currently have a 'click' sonifier that applies to all event-based data, for example, beats.  We also have two beat namespaces, one with metrical position and one without.  This conversion can go in only one direction: `beat_position -> beat`.

Now, imagine that we want to have a sonifier that plays different clicks on the downbeat.  This would only apply to the `beat_position` namespace, but that can also be coerced into a `beat` namespace.

Proposed solution: use an ordered dict to store the mappings, and put the most specific namespaces first.  The conversion logic stays the same, but sonification is always done with the most specific applicable method.  If you want to sonify using an alternate method, you can always auto-convert the annotation before sonifying.

How does this sound?